### DISCUSSION TITLE
Please create a non-composer way of installing

### DISCUSSION THREAD

**Discussion by ppp-x** [PHPOffice/PhpSpreadsheet#3296] [Ideas]
Composer is not always available for people, for example due to server restrictions on shared servers. Requiring composer makes this software unusable for some people. Please provide another way of installing.

And yes, I know it's possible to run composer locally and then upload the files to the server but this eliminates the benefit of composer anyway and basically is just an extra step to achieve something that could be achieved easier by making other installation options available, not to mention that a local composer may think certain dependencies are met that are on the local system but aren't on the server.

**Comment by MarkBaker** [comment]
[Please Read](https://github.com/PHPOffice/PhpSpreadsheet/discussions/1969)

 > And yes, I know it's possible to run composer locally and then upload the files to the server but this eliminates the benefit of composer anyway

Not really; because you're still getting all the autoloading and namespacing handled cleanly then; when you install/build your app using composer, it handles all this for you even though it may not be installed on the server.


 > and basically is just an extra step to achieve something that could be achieved easier by making other installation options available,

No it isn't! It does not eliminate a lot of steps! It doesn't eliminate steps for us as maintainers; what might be one simple step extra step for you is a lot of extra steps for us. You're asking us to provide all the dependencies, scope everything for namespacing in case there are version conflicts with any other libraries that you might have, bundle everything up into a single package (including optional libraries?!?), provide and maintain our own autoloader, test the builds, etc.
I'm sure you'd complain about a bloated library if we included all the optional dependencies like mitoteam/jpgraph, mpdf/mpdf
dompdf/dompdf, and tecnickcom/tcpdf; or missing functionality if we didn't include those.
We'd also need to maintain different bundles (with different dependency versions) for the range of PHP versions that we support (currently from 7.4 to 8.2)

It was necessary to do this in the days before composer, and it took a full 8-10 hours of work building every release, and that was with automated scripts that I ran to create those builds.

> not to mention that a local composer may think certain dependencies are met that are on the local system but aren't on the server.

That's why the your composer.json on your local dev box should have a `config` section that defines the server environment if it doesn't match your development environment.

**Reply by ppp-x** [reply]
Your "please read" just links to a hars, unnecessarily rude post with no help at all. You already mention you had scripts that automated the making of these builds, so I would assume the 8-10 hours of work is simply waiting until it's done. Perhaps you have stocks in composer or sth but in my experience it's absolute worthless rubbish. Editing the `config` section in composer.json assumes all the necessary information is available to the developer, which again, isn't always the case if using a shared server with many restrictions and even if done correctly, is certainly not a reliable workaround and half the time simply won't do the right thing. Requiring composer really makes installation impossible for many people.

**Reply by MarkBaker** [reply]
> Your "please read" just links to a hars, unnecessarily rude post with no help at all.

My "Please Read" links to a post that suggests running composer on your developer box; and that also explains the basic steps that you should take to build an install manually (without composer). It is informational; not harsh. If anybody is being rude here, it's you.

> You already mention you had scripts that automated the making of these builds, so I would assume the 8-10 hours of work is simply waiting until it's done.

I **_had_** scripts 15 years ago, back in the days when this library was PHPExcel: I don't have them any more (and even if I did, they wouldn't work directly with PhpSpreadsheet) because they are not needed when Composer can be used instead. But then, you're not talking about something that would cost you any time, it would be my time that it would take. Clearly you value my time less than you value your own.

> Perhaps you have stocks in composer or sth but in my experience it's absolute worthless rubbish.

Composer is a free tool, so having stocks in Composer would be meaningless, a guaranteed 0% return.
As for Composer being "absolute worthless rubbish", it seems that the vast majority of PHP developers disagree with you. Composer has become the de facto standard for dependency management in the PHP world, and is used by almost all libraries, tools and projects.

> Editing the config section in composer.json assumes all the necessary information is available to the developer, which again, isn't always the case if using a shared server with many restrictions and even if done correctly, is certainly not a reliable workaround and half the time simply won't do the right thing

If a developer doesn't know the PHP version and modules that are installed for their target server, then they have a lot more serious problems. Even a basic `phpinfo()` script should provide that information.

> Requiring composer really makes installation impossible for many people.

No, it makes it more difficult for a very few people. Most developers work with composer quite happily.
